### PR TITLE
size_report: put symbols in a tree structure using anytree

### PR DESCRIPTION
**Pull request by nashif**
- Use anytree module to display tree and to allow easy exporting into json.
- Add option to export results into a json file.
- Generate json files when target is call and put them in cmake binary directory

### CODE REVIEW

**Comment by carlescufi** [comment]
Looks very nice!

Fails on Windows:

```
FAILED: zephyr/cmake/reports/CMakeFiles/rom_report
cmd.exe /C "cd /D C:\Users\Carles\src\zephyr\zephyr\build\zephyr\cmake\reports && C:\Python37\python.exe C:/Users/Carles/src/zephyr/zephyr/scripts/footprint/size_report -k C:/Users/Carles/src/zephyr/zephyr/build/zephyr/zephyr.elf -z C:/Users/Carles/src/zephyr/zephyr -o C:/Users/Carles/src/zephyr/zephyr/build --json C:/Users/Carles/src/zephyr/zephyr/build/rom.json rom"
Traceback (most recent call last):
  File "C:/Users/Carles/src/zephyr/zephyr/scripts/footprint/size_report", line 623, in <module>
    main()
  File "C:/Users/Carles/src/zephyr/zephyr/scripts/footprint/size_report", line 604, in main
    do_address_range_matching(elf, symbol_dict, processed)
  File "C:/Users/Carles/src/zephyr/zephyr/scripts/footprint/size_report", line 417, in do_address_range_matching
    path = get_die_filename(die, lineprog)
  File "C:/Users/Carles/src/zephyr/zephyr/scripts/footprint/size_report", line 244, in get_die_filename
    path = path.resolve()
  File "C:\Python37\lib\pathlib.py", line 1161, in resolve
    s = self._flavour.resolve(self, strict=strict)
  File "C:\Python37\lib\pathlib.py", line 205, in resolve
    s = self._ext_to_normal(_getfinalpathname(s))
OSError: [WinError 123] The filename, directory name, or volume label syntax is incorrect: 'C:\\Users\\Carles\\src\\zephyr\\zephyr\\build\\<built-in>'
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: 'C:\Program Files\CMake\bin\cmake.EXE' --build 'C:\Users\Carles\src\zephyr\zephyr\build' --target rom_report
```

**Comment by nashif** [comment]
> Fails on Windows (also in master actually):

@carlescufi  can you try with the patch added? guessing this might be the problem.

**Comment by cfriedt** [comment]
Cool - does this print a tree of all of the symbols in the address space?

File: scripts/footprint/size_report [hunk 1]
```diff
@@ -495,89 +476,101 @@ def set_root_path_for_unmapped_symbols(symbol_dict, addr_range, processed):
...
+        if not row.node.children:
+            cc = Fore.CYAN
+            cr = Fore.RESET
+        elif row.node.name.endswith(".c") or row.node.name.endswith(".h"):
+            cc = Fore.GREEN
+            cr = Fore.RESET
+
+        print(f"{row.pre}{cc}{row.node.name}{cr} {s} {Fore.BLUE}{percent:.2f}%{Fore.RESET}")
+    print('=' * 110)
+    print(f'{total_size:>100}')
```
Comment (dcpleung) [inline scripts/footprint/size_report:552]: Probably needs to be `>101`. It is misaligned when I ran it.

**Review by dcpleung — COMMENTED** [review]
Wonder if the nodes can be sorted in alphabetical order... I ran hello_world with qemu_x86, and the files under `lib/os` are not sorted (function names are not either).

File: scripts/footprint/size_report (hunk 1 above)
Comment (nashif) [reply scripts/footprint/size_report]: fixed

**Comment by nashif** [comment]
> Wonder if the nodes can be sorted in alphabetical order...

done

**Review by dcpleung — APPROVED** [review]

**Comment by dcpleung** [comment]
> Cool - does this print a tree of all of the symbols in the address space?

This should print all visible symbols in the ELF file.

**Review by carlescufi — COMMENTED** [review]
Shouldn't you add `anytree` to the requirements files?

**Comment by nashif** [comment]
> Shouldn't you add `anytree` to the requirements files?

done

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**